### ISSUE TITLE
Phalcon Double Sends Headers & Cookies

### ISSUE THREAD

**Issue by Bellardia** [phalcon/cphalcon#12908]
### Expected Behavior
Cookies and Headers get output once per request lifecycle.

### Actual Behavior
Cookies and Headers set in `\Phalcon\Http\Response` will get output twice if a controller returns an object implementing `ResponseInterface`.

`cphalcon/phalcon/mvc/application.zep` inside `handle`
```zep
if typeof possibleResponse == "object" {
	if possibleResponse instanceof ResponseInterface {
		possibleResponse->sendHeaders();
		possibleResponse->sendCookies();
		return possibleResponse;
	}
}
```

`cphalcon/phalcon/http/response.zep`
```zep
public function send() -> <Response>
	{
		var content, file;

		if this->_sent {
			throw new Exception("Response was already sent");
		}

		this->sendHeaders();

		this->sendCookies();

		/**
		 * Output the response body
		 */
		let content = this->_content;
		if content != null {
			echo content;
		} else {
			let file = this->_file;

			if typeof file == "string" && strlen(file) {
				readfile(file);
			}
		}

		let this->_sent = true;
		return this;
	}
```

```http
Set-Cookie:test=test; Max-Age=0; path=/; domain=.localhost
Set-Cookie:test=test; Max-Age=0; path=/; domain=.localhost
```

**Comment by Jurigag** [comment]
Just don't use send method if you return response? Application doesn't call send anywhere so not sure what is your problem.

**Comment by Bellardia** [comment]
@Jurigag 

Are you using Micro or MVC?

The send() method is not called automatically in all configurations, which may mean while it works for you it is still not correct. 

`cphalcon/phalcon/mvc/micro.zep`
```
/**
 * Check if the returned object is already a response
 */
if typeof returnedValue == "object" {
	if returnedValue instanceof ResponseInterface {
		/**
		 * Automatically send the response
		 */
		 if !returnedValue->isSent() {
		 	returnedValue->send();
		 }
	}
}
```

Here we can see micro correctly calls the `send()` method on the response, but this is not done in MVC.

`cphalcon/phalcon/http/response.zep`
```
if typeof possibleResponse == "object" {
	if possibleResponse instanceof ResponseInterface {
		possibleResponse->sendHeaders();
		possibleResponse->sendCookies();
		return possibleResponse;
	}
}
```

As `Phalcon\Http\Response::send()` outputs headers, you cannot avoid the double send because `Phalcon\Mvc\Application::handle()` will also send headers.

**Comment by Jurigag** [comment]
I am using MVC, but what micro has to do with application? It's different kernel, you shouldn't use both at the same time. So i still don't know what's your issue.

**Comment by Bellardia** [comment]
I'm not using Micro, it was an inquiry. 

I'm not sure where we differ here but if you follow the code path there's an issue. I appreciate your taking the time to respond, but you're not providing any productive feedback.

**Comment by Jurigag** [comment]
I'm not using too and i don't have this issue.

**Comment by Bellardia** [comment]
With all do respect you're not being helpful.

`getContent()` is not the same function as `send()`. The question is as to why `handle()` decides to send headers but not output makes little sense from an API perspective. 

`getContent()` does not update class flags such as `_sent`. Which again, leads me to believe this is a bug.

**Comment by grigoriygithub** [comment]
I found PR #12378 https://github.com/phalcon/cphalcon/pull/12378/commits/ac254c9fbfca27e5bc9571353be87a216fc1d916

**Comment by Bellardia** [comment]
@grigoriygithub Amazing, that's exactly the issue :) Thanks!!

That diff is quite unwieldily at this time though.

**Comment by Jurigag** [comment]
Yea, PR is fine i guess, just not sure if we can add it in phalcon 3.

**Comment by TheCrimsonIdol** [comment]
??? this should not be.

**Comment by sergeyklay** [comment]
Fixed in the 4.0.x branch
